### PR TITLE
Added more detailed logs and generating k8s events

### LINKED ISSUE #52: Improve logging and add events
The `rcs-ocm-deployer` code currently includes some very basic logging and this should be enriched so that the code is easier to debug.

### PR DESCRIPTION
**Pull request by sahar2339**
Changed logs to be more detailed and chained together for easier debugging and tracebacks. Also added Kubernetes events to be generated for major successful or failed operations.

fixes #52

### CODE REVIEW

File: internals/controllers/cappPlacementSync_controller.go [hunk 1]
```diff
@@ -76,37 +77,40 @@ var CappPredicateFuncs = predicate.Funcs{
...
 		if errors.IsNotFound(err) {
 			mw := utils.GenerateManifestWorkGeneric(mwName, managedClusterName, manifests, workv1.ManifestConfigOption{})
 			utils.SetManifestWorkCappAnnotations(*mw, capp)
 			if err := r.Create(ctx, mw); err != nil {
-				return ctrl.Result{}, err
+				r.EventRecorder.Event(&capp, "Error", "FailedToCreateManifestWork", err.Error())
+				return ctrl.Result{}, fmt.Errorf("Failed to create ManifestWork %s", err.Error())
 			}
+			logger.Info(fmt.Sprintf("Created ManifestWork %s for capp %s", mwName, capp.Name))
+			r.EventRecorder.Event(&capp, "Normal", "CreatedManifestWork", fmt.Sprintf("Created ManifestWork %s for capp %s", mwName, capp.Name))
```
Comment (mzeevi) [inline internals/controllers/cappPlacementSync_controller.go:100]: These should all be constants instead of strings... `Normal`, `CreatedManifestWork` etc

Comment (mzeevi) [reply internals/controllers/cappPlacementSync_controller.go:100]: See here for inspiration https://github.com/medik8s/node-healthcheck-operator/blob/main/controllers/nodehealthcheck_controller.go#L69

File: internals/controllers/cappPlacementSync_controller.go [hunk 2]
```diff
@@ -76,37 +77,40 @@ var CappPredicateFuncs = predicate.Funcs{
...
 			return ctrl.Result{}, nil
 		}
 		return ctrl.Result{}, err
 	}
 	mw.Spec.Workload.Manifests = manifests
 
 	if err = r.Update(ctx, &mw); err != nil {
-		l.Error(err, "unable to update ManifestWork")
 		if errors.IsConflict(err) {
+			logger.Info(fmt.Sprint("Conflict while updating ManifestWork trying again in a few seconds"))
```
Comment (mzeevi) [inline internals/controllers/cappPlacementSync_controller.go:109]: Add a comma before the word trying

File: internals/controllers/cappPlacementSync_controller.go [hunk 3]
```diff
@@ -76,37 +77,40 @@ var CappPredicateFuncs = predicate.Funcs{
...
 			return ctrl.Result{RequeueAfter: time.Second * 2}, nil
 		}
-		return ctrl.Result{}, err
+		return ctrl.Result{}, fmt.Errorf("Failed to sync ManifestWork %s", err.Error())
```
Comment (mzeevi) [inline internals/controllers/cappPlacementSync_controller.go:112]: If you concatenate errors then it'd be more readable to do it after a `:`, like this 

```
fmt.Errorf("Failed to sync ManifestWork: %s", err.Error()
```

File: internals/controllers/cappPlacement_controller.go [hunk 4]
```diff
@@ -44,19 +48,22 @@ func (r *ServicePlacementReconciler) Reconcile(ctx context.Context, req ctrl.Req
...
+			logger.Info(fmt.Sprintf("requeuing capp %s, waiting for PlacementDecision to be satisfied.", capp.Name))
+			r.EventRecorder.Event(&capp, "Warning", "PlacementDecisionNotSatisfied", fmt.Sprintf("Failed to schedule capp %s to managed cluster. PlacementDecision with optional clusters was not found for placement %s.", capp.Name, placementRef))
 			return ctrl.Result{RequeueAfter: 10 * time.Second * 2}, nil
 		}
 		placementRef = cluster
 	}
 	if err := utils.UpdateCappDestination(capp, placementRef, ctx, r.Client); err != nil {
-		l.Error(err, "unable to update capp")
-		return ctrl.Result{}, err
+		return ctrl.Result{}, fmt.Errorf("Unable to update capp with selected cluster %s", err.Error())
```
Comment (mzeevi) [inline internals/controllers/cappPlacement_controller.go:64]: There's a lack of consistency: at times it's written as `capp` and at times as `Capp`. I think we should go with `Capp` all across the logs and events

File: internals/controllers/cappPlacement_controller.go [hunk 5]
```diff
@@ -44,19 +48,22 @@ func (r *ServicePlacementReconciler) Reconcile(ctx context.Context, req ctrl.Req
...
 	}
+	r.EventRecorder.Event(&capp, "Normal", "CappScheduled", fmt.Sprintf("Scheduled Capp %s for managed cluster %s", capp.Name, placementRef))
```
Comment (mzeevi) [inline internals/controllers/cappPlacement_controller.go:66]: Scheduled on, not for*

File: internals/controllers/cappPlacement_controller.go [hunk 6]
```diff
@@ -97,20 +104,18 @@ func (r *ServicePlacementReconciler) pickDecision(capp rcsv1alpha1.Capp, log log
 	}
 	placement := clusterv1beta1.Placement{}
 	if err := r.Client.Get(ctx, types.NamespacedName{Name: placementRef, Namespace: r.PlacementsNamespace}, &placement); err != nil {
-		return "", err
+		return "", fmt.Errorf("Failed to get placement %s", err.Error())
 	}
 	placementDecisions, err := utils.GetPlacementDecisionList(capp, log, ctx, placementRef, r.PlacementsNamespace, r.Client)
 	if len(placementDecisions.Items) == 0 {
-		log.Info("unable to find any PlacementDecision, try again after 10 seconds")
```
Comment (mzeevi) [inline internals/controllers/cappPlacement_controller.go:104]: Was this not an indicative log message?

File: internals/utils/capp.go [hunk 7]
```diff
@@ -82,7 +83,7 @@ func GatherCappResources(capp rcsv1alpha1.Capp, ctx context.Context, l logr.Logg
 func UpdateCappDestination(capp rcsv1alpha1.Capp, managedClusterName string, ctx context.Context, r client.Client) error {
 	capp.Status.ApplicationLinks.Site = managedClusterName
 	if err := r.Status().Update(ctx, &capp); err != nil {
-		return err
+		return fmt.Errorf("Failed to update capp status with selected site %s", err.Error())
```
Comment (mzeevi) [inline internals/utils/capp.go:86]: More readable as:

```
fmt.Errorf("Failed to update capp status with selected site: %s", err.Error())
```

File: internals/utils/capp.go [hunk 8]
```diff
@@ -107,7 +108,7 @@ func RemoveCreatedAnnotation(ctx context.Context, service rcsv1alpha1.Capp, r cl
 	delete(cappAnno, "AnnotationNamespaceCreated")
 	service.SetAnnotations(cappAnno)
 	if err := r.Update(ctx, &service); err != nil {
-		return err
+		return fmt.Errorf("Failed to remove annotations from capp %s", err.Error())
```
Comment (mzeevi) [inline internals/utils/capp.go:111]: More readable as:
```
fmt.Errorf("Failed to update Capp status with selected site: %s", err.Error())
```

File: internals/utils/finalizer.go [hunk 9]
```diff
@@ -60,7 +59,7 @@ func EnsureFinalizer(ctx context.Context, service rcsv1alpha1.Capp, r client.Cli
 	if !controllerutil.ContainsFinalizer(&service, FinalizerCleanupCapp) {
 		controllerutil.AddFinalizer(&service, FinalizerCleanupCapp)
 		if err := r.Update(ctx, &service); err != nil {
-			return err
+			return fmt.Errorf("Failed to add finalizer to capp %s", err.Error())
```
Comment (mzeevi) [inline internals/utils/finalizer.go:62]: Change to (also in previous places):

```
fmt.Errorf("Failed to update Capp status with selected site: %s", err.Error())
```

File: internals/utils/placement.go [hunk 10]
```diff
@@ -48,7 +47,7 @@ func GetDecisionClusterName(placementDecisions *clusterv1beta1.PlacementDecision
 		managedClusterName = pd.Status.Decisions[1].ClusterName
 	}
 	if len(managedClusterName) == 0 {
-		log.Info("unable to find a valid ManagedCluster from PlacementDecision, try again after 10 seconds")
+		log.Info("Unable to find a valid ManagedCluster from PlacementDecision.")
```
Comment (mzeevi) [inline internals/utils/placement.go:50]: Log messages shouldn't end in a dot (`.`)

File: internals/utils/status/conditions.go [hunk 11]
```diff
@@ -61,7 +61,7 @@ func generateCondtion(condType string, status bool, reason ...string) metav1.Con
 func SetVolumesCondition(capp rcsv1alpha1.Capp, ctx context.Context, r client.Client, l logr.Logger, status bool, reason ...string) error {
 	condition := generateCondtion(condVolumesType, status, reason...)
 	if err := setCappCondition(capp, ctx, r, l, condition); err != nil {
-		return err
+		return fmt.Errorf("Failed to set volumeCondition to capp %s", err)
```
Comment (mzeevi) [inline internals/utils/status/conditions.go:64]: `on` instead of `to`

File: internals/controllers/cappPlacementSync_controller.go [hunk 12]
```diff
@@ -76,37 +77,40 @@ var CappPredicateFuncs = predicate.Funcs{
...
+func (r *ServiceNamespaceReconciler) SyncManifestWork(capp rcsv1alpha1.Capp, ctx context.Context, logger logr.Logger) (ctrl.Result, error) {
 	mwName := utils.NamespaceManifestWorkPrefix + capp.Namespace + "-" + capp.Name
 	managedClusterName := capp.Annotations[utils.AnnotationKeyHasPlacement]
 	var mw workv1.ManifestWork
-	manifests, err := utils.GatherCappResources(capp, ctx, l, r.Client)
+	manifests, err := utils.GatherCappResources(capp, ctx, logger, r.Client)
 	if err != nil {
-		statusutils.SetVolumesCondition(capp, ctx, r.Client, l, false, err.Error())
-		return ctrl.Result{}, err
+		r.EventRecorder.Event(&capp, "Error", "VolumeWasNotFound", err.Error())
```
Comment (mzeevi) [inline internals/controllers/cappPlacementSync_controller.go:86]: I think `VolumeNotFound` is a more conventional name than `VolumeWasNotFound`

File: internals/controllers/cappPlacementSync_controller.go [hunk 13]
```diff
@@ -76,37 +77,40 @@ var CappPredicateFuncs = predicate.Funcs{
...
+		return ctrl.Result{}, fmt.Errorf("Failed to get one of the volumes from capp spec %s", err.Error())
 	}
 
 	if err := r.Get(ctx, types.NamespacedName{Name: mwName, Namespace: managedClusterName}, &mw); err != nil {
 		if errors.IsNotFound(err) {
 			mw := utils.GenerateManifestWorkGeneric(mwName, managedClusterName, manifests, workv1.ManifestConfigOption{})
 			utils.SetManifestWorkCappAnnotations(*mw, capp)
 			if err := r.Create(ctx, mw); err != nil {
-				return ctrl.Result{}, err
+				r.EventRecorder.Event(&capp, "Error", "FailedToCreateManifestWork", err.Error())
```
Comment (mzeevi) [inline internals/controllers/cappPlacementSync_controller.go:96]: Think a better naming would be `ManifestWorkCreationFailed`

File: internals/controllers/cappPlacementSync_controller.go (hunk 1 above)
Comment (mzeevi) [reply internals/controllers/cappPlacementSync_controller.go:100]: A better name would be `ManifestWorkCreated`

File: internals/controllers/cappPlacement_controller.go (hunk 6 above)
Comment (sahar2339) [reply internals/controllers/cappPlacement_controller.go:104]: duplicate

File: internals/controllers/cappPlacementSync_controller.go [hunk 14]
```diff
@@ -35,26 +38,24 @@ const (
...
 		return ctrl.Result{}, err
 	}
-
-	l.Info("Start reconciling Service in ServiceNamespace controller")
-	if err, deleted := utils.HandleResourceDeletion(ctx, capp, l, r.Client); err != nil {
-		if deleted {
-			return ctrl.Result{}, nil
+	if capp.ObjectMeta.DeletionTimestamp != nil {
+		if err := utils.HandleResourceDeletion(ctx, capp, logger, r.Client); err != nil {
+			return ctrl.Result{}, err
```
Comment (mzeevi) [inline internals/controllers/cappPlacementSync_controller.go:51]: This code is from #56, no?

**Comment by mzeevi** [comment]
Convention should be:

1. `Capp` instead of `capp` in logs and events.
2. Error messages (`fmt.Errorf`) should start in lower case
3. Info messages (`logger.Info`) should start in upper case
4. Event messages (`r.EventRecorder.Event`) should start in upper case

File: internals/controllers/cappPlacementSync_controller.go [hunk 15]
```diff
@@ -76,37 +77,40 @@ var CappPredicateFuncs = predicate.Funcs{
...
 	mw.Spec.Workload.Manifests = manifests
 
 	if err = r.Update(ctx, &mw); err != nil {
-		l.Error(err, "unable to update ManifestWork")
 		if errors.IsConflict(err) {
+			logger.Info(fmt.Sprint("conflict while updating manifestWork trying again in a few seconds"))
 			return ctrl.Result{RequeueAfter: time.Second * 2}, nil
 		}
-		return ctrl.Result{}, err
+		return ctrl.Result{}, fmt.Errorf("failed to sync manifestWork: %s", err.Error())
```
Comment (mzeevi) [inline internals/controllers/cappPlacementSync_controller.go:112]: Be consistent with the naming, there are other places where it's written as `ManifestWork` and not `manifestWork`

File: internals/utils/placement.go [hunk 16]
```diff
@@ -39,7 +38,7 @@ func GetPlacementDecisionList(capp rcsv1alpha1.Capp, log logr.Logger, ctx contex
 func GetDecisionClusterName(placementDecisions *clusterv1beta1.PlacementDecisionList, log logr.Logger) string {
 	pd := placementDecisions.Items[0]
 	if len(pd.Status.Decisions) == 0 {
-		log.Info("unable to find any Decisions from PlacementDecision, try again after 10 seconds")
+		log.Info("unable to find PlacementDecision.")
```
Comment (mzeevi) [inline internals/utils/placement.go:41]: remove `.`

File: internals/utils/placement.go [hunk 17]
```diff
@@ -48,7 +47,7 @@ func GetDecisionClusterName(placementDecisions *clusterv1beta1.PlacementDecision
 		managedClusterName = pd.Status.Decisions[1].ClusterName
 	}
 	if len(managedClusterName) == 0 {
-		log.Info("unable to find a valid ManagedCluster from PlacementDecision, try again after 10 seconds")
+		log.Info("unable to find a valid ManagedCluster from PlacementDecision.")
```
Comment (mzeevi) [inline internals/utils/placement.go:50]: remove `.`

File: internals/utils/volumes.go [hunk 18]
```diff
@@ -56,6 +55,7 @@ func prepareVolumesManifests(secrets []string, configMaps []string, capp rcsv1al
 					Namespace: secret.Namespace,
 				},
 				Data: secret.Data,
+				Type: secret.Type,
```
Comment (mzeevi) [inline internals/utils/volumes.go:58]: This code is part of #56 and shouldn't actually be part of this PR

**Review by mzeevi — APPROVED** [review]

**Pull request merged**